File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // 2 pointer at distance n
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode *slow = dummy, *fast = dummy;
        int cnt = 0;
        while(fast){
            cnt++;
            if(cnt>n+1) slow = slow->next;
            fast=fast->next;
        }
        ListNode * nextNode = slow->next;
        slow->next = nextNode?nextNode->next:NULL;
        return dummy->next;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        l1,l2,prev = head,head,None 

        while l2.next: 
            l2=l2.next
            if n<=0:
                l1=l1.next 
            n-=1 
        
        if n>0:
            return head.next 
        l1.next =l1.next.next 
        return head

